web-common/src/features/entity-management/file-artifacts-store.ts

```diff
@@ -1,131 +1,241 @@
-) => {
-  update((state) => {
-    if (!state.entities[path]) {
-      state.entities[path] = {
-        errors: [],
-        jsonRepresentation: {},
-        isReconciling: false,
-      };
+import { removeLeadingSlash } from "@rilldata/web-common/features/entity-management/entity-mappers";
+import {
+  ResourceKind,
+  useProjectParser,
+  useResource,
+} from "@rilldata/web-common/features/entity-management/resource-selectors";
+import { getSubStore } from "@rilldata/web-common/lib/getSubStore";
+import {
+  runtimeServiceListResources,
+  type V1ParseError,
+  V1ReconcileStatus,
+  type V1Resource,
+  type V1ResourceName,
+} from "@rilldata/web-common/runtime-client";
+import type { QueryClient } from "@tanstack/svelte-query";
+import { derived, get, type Readable, writable } from "svelte/store";
+
+export class FileArtifact {
+  public readonly path: string;
+
+  public readonly name = writable<V1ResourceName | undefined>(undefined);
+
+  /**
+   * Last time the state of the resource `kind/name` was updated
+   * used to make sure we do not have unnecessary refreshes
+   */
+  public lastStateUpdatedOn: string | undefined;
+
+  public constructor(filePath: string) {
+    this.path = filePath;
+  }
+
+  public updateResource(resource: V1Resource) {
+    const curName = get(this.name);
+    if (
+      curName?.name !== resource.meta?.name?.name ||
+      curName?.kind !== resource.meta?.name?.kind
+    ) {
+      this.name.set({
+        kind: resource.meta?.name?.kind,
+        name: resource.meta?.name?.name,
+      });
+    }
+    this.lastStateUpdatedOn = resource.meta?.stateUpdatedOn;
+  }
+
+  public deleteResource() {
+    this.name.set(undefined);
+  }
+}
+
+export class FileArtifactsStore {
+  /**
+   * Map of all files and whether it is reconciling or not.
+   * If an entry is present here then there should be on in {@link artifacts} as well
+   */
+  public readonly files = writable<Record<string, boolean>>({});
+
+  /**
+   * Map of all files and its individual store
+   */
+  private readonly artifacts: Record<string, FileArtifact> = {};
+
+  // Actions
+
+  public async init(instanceId: string) {
+    const resourcesResp = await runtimeServiceListResources(instanceId);
+    if (!resourcesResp.resources?.length) return;
+    for (const resource of resourcesResp.resources) {
+      switch (resource.meta?.name?.kind) {
+        case ResourceKind.Source:
+        case ResourceKind.Model:
+        case ResourceKind.MetricsView:
+        case ResourceKind.Chart:
+        case ResourceKind.Dashboard:
+          this.setResource(resource);
+          break;
+      }
     }
-    callback(state.entities[path]);
-    return state;
-  });
-};
-
-const fileArtifactsEntitiesReducers = {
-  setName(path: string, name: string) {
-    createOrUpdateFileArtifact(path, (entityData) => {
-      entityData.name = name;
+  }
+
+  public updateFile(filePath: string) {
+    if (get(this.files)[filePath]) return;
```

Comment:
We would need to make sure to add `FileArtifact` for a path. This method handles that. Since changes to file are returned in a different watch request this is needed.